### PR TITLE
Adding a state and do not delete multisig transactions when they're executed/expired/withdrawn

### PR DESCRIPTION
**Pull request by hansl**
This requires https://github.com/liftedinit/many-rs/pull/114

### CODE REVIEW

**Comment by hansl** [comment]
I added a test and a bunch of mechanics to work with blocks in testing. I will add more tests tomorrow for the edge cases.

File: src/many-ledger/src/storage.rs [hunk 1]
```diff
@@ -469,30 +481,41 @@ impl LedgerStorage {
         use rocksdb::{Direction, IteratorMode, ReadOptions};
 
         let mut options = ReadOptions::default();
-        let mut bound = TRANSACTIONS_ROOT.to_vec();
+        options.set_iterate_lower_bound(MULTISIG_TRANSACTIONS_ROOT);
+        let mut bound = MULTISIG_TRANSACTIONS_ROOT.to_vec();
         bound[MULTISIG_TRANSACTIONS_ROOT.len() - 1] += 1;
```
Comment (fmorency) [inline src/many-ledger/src/storage.rs:486]: What is this doing exactly?

File: src/many-ledger/src/storage.rs [hunk 2]
```diff
@@ -469,30 +481,41 @@ impl LedgerStorage {
...
+            if storage.info.timeout.0 > now {
+                if !storage.disabled {
+                    storage.disabled = true;
+                    storage.info.state = MultisigTransactionState::Expired;
+
+                    if let Ok(v) = minicbor::to_vec(storage) {
+                        batch.push((k.to_vec(), Op::Put(v)));
+                    }
+                }
+            } else if let Ok(d) = now.duration_since(storage.creation) {
```
Comment (fmorency) [inline src/many-ledger/src/storage.rs:510]: Since the DB is ordered, at this point we don't need to continue since we know that the rest is all timed out anyway.

**Review by fmorency — CHANGES_REQUESTED** [review]
Looking good. Some comments. Thanks! I'll let you add more tests.

File: src/many-ledger/src/storage.rs (hunk 2 above)
Comment (hansl) [reply src/many-ledger/src/storage.rs:510]: I assume you want me to add a comment?  ^_^

Comment (fmorency) [reply src/many-ledger/src/storage.rs:510]: If you want; it was mostly for my comprehension.

File: src/many-ledger/src/storage.rs [hunk 3]
```diff
@@ -1005,7 +1028,11 @@ impl LedgerStorage {
                 execute_automatically,
                 timeout,
                 data: arg.data.clone(),
+
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-ledger/src/storage.rs:1031]: rustfmt fail...

File: src/many-ledger/src/storage.rs [hunk 4]
```diff
@@ -469,30 +481,41 @@ impl LedgerStorage {
...
 
-            let storage: MultisigTransactionStorage = minicbor::decode(v.value())
+            let mut storage: MultisigTransactionStorage = minicbor::decode(v.value())
                 .map_err(|e| ManyError::deserialization_error(e.to_string()))?;
+            let now = self.now();
 
-            if storage.info.timeout.0 > self.current_time.unwrap_or_else(SystemTime::now) {
-                keys_to_delete.push(k.to_vec());
+            if storage.info.timeout.0 > now {
+                if !storage.disabled {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-ledger/src/storage.rs:502]: We don't have the transaction ID in this position, we would need to decode it from the key. At that point it's more work to DRY than to RY.

[the remark this answers is not in the document]
Comment (hansl) [reply src/many-ledger/src/storage.rs:502]: I did move some logic into the Storage class which makes it less copy-pasted, at least.

File: src/many-ledger/tests/block.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+//! Tests regarding blockchain behaviour.
+#![allow(dead_code)]
+mod common;
```
Comment (fmorency) [inline src/many-ledger/tests/block.rs:3]: I think you can use `pub mod common;` and remove the `#![allow(dead_code)]`

File: src/many-ledger/tests/common/mod.rs [hunk 6]
```diff
@@ -99,18 +118,65 @@ impl Setup {
...
+                BalanceArgs {
+                    account: None,
+                    symbols: Some(vec![symbol].into()),
+                },
+            )?
+            .balances
+            .get(&symbol)
+            .cloned()
+            .unwrap_or_default())
+    }
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:135]: nit: empty line

File: src/many-ledger/tests/common/mod.rs [hunk 7]
```diff
@@ -126,6 +192,97 @@ impl Setup {
...
+                    memo: Some("Foo".to_string()),
+                    transaction: Box::new(transaction),
+                    threshold: None,
+                    timeout_in_secs: None,
+                    execute_automatically: None,
+                    data: None,
+                },
+            )
+            .map(|x| x.token)
+    }
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:216]: nit: empty line

File: src/many-ledger/tests/common/mod.rs [hunk 8]
```diff
@@ -126,6 +192,97 @@ impl Setup {
...
+            many::types::ledger::AccountMultisigTransaction::Send(
+                many::server::module::ledger::SendArgs {
+                    from: Some(account_id),
+                    to,
+                    symbol,
+                    amount: amount.into(),
+                },
+            ),
+        )
+    }
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:244]: nit: empty line

File: src/many-ledger/tests/common/mod.rs [hunk 9]
```diff
@@ -126,6 +192,97 @@ impl Setup {
...
+
+    /// Approve a multisig transaction.
+    pub fn multisig_approve(&mut self, id: Identity, token: &ByteVec) -> Result<(), ManyError> {
+        let token = token.clone();
+        self.module_impl
+            .multisig_approve(&id, account::features::multisig::ApproveArgs { token })?;
+        Ok(())
+    }
+    pub fn multisig_approve_(&mut self, id: Identity, token: &ByteVec) {
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:262]: nit: empty line

File: src/many-ledger/tests/common/mod.rs [hunk 10]
```diff
@@ -126,6 +192,97 @@ impl Setup {
...
+            .expect("Could not approve multisig")
+    }
+
+    /// Execute the transaction.
+    pub fn multisig_execute(&mut self, token: &ByteVec) -> Result<ResponseMessage, ManyError> {
+        let token = token.clone();
+        self.module_impl
+            .multisig_execute(&self.id, ExecuteArgs { token })
+    }
+    pub fn multisig_execute_(&mut self, token: &ByteVec) -> ResponseMessage {
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:273]: nit: empty line

File: src/many-ledger/tests/multisig.rs [hunk 11]
```diff
@@ -474,65 +474,118 @@ fn withdraw_invalid() {
...
+    assert_eq!(h, 2);
+    tokens.append(&mut tokens2);
+
+    // Approve 4 of them in a block. Execute 2.
+    let (h, _) = setup.block(|setup| {
+        setup.multisig_approve_(identity(2), &tokens[0]);
+        setup.multisig_approve_(identity(2), &tokens[1]);
+        setup.multisig_approve_(identity(2), &tokens[2]);
+        setup.multisig_approve_(identity(2), &tokens[3]);
         assert_eq!(
```
Comment (fmorency) [inline src/many-ledger/tests/multisig.rs:551]: Use `assert_many_err()`?

File: src/many-ledger/tests/multisig.rs [hunk 12]
```diff
@@ -474,65 +474,118 @@ fn withdraw_invalid() {
...
+        setup.multisig_approve_(identity(3), &tokens[2]);
+        setup.multisig_approve_(identity(3), &tokens[3]);
+
+        // Is okay.
+        setup.send_(setup.id, account_ids[2], 100u32);
+        let data = setup.multisig_execute_(&tokens[2]).data;
+        assert!(data.is_ok(), "Err: {}", data.unwrap_err());
+
+        // Insufficient funds.
+        assert!(setup.multisig_execute_(&tokens[3]).data.is_err());
```
Comment (fmorency) [inline src/many-ledger/tests/multisig.rs:565]: nit: check the error

File: src/many-ledger/tests/common/mod.rs [hunk 13]
```diff
@@ -99,18 +118,65 @@ impl Setup {
...
+        self.send(from, to, amount, *MFX_SYMBOL)
+            .expect("Could not send tokens")
+    }
+
+    pub fn create_account(&mut self, account_type: AccountType) -> Result<Identity, ManyError> {
         let args = self.create_account_args(account_type);
-        self.module_impl.create(&self.id, args).unwrap().id
+        self.module_impl.create(&self.id, args).map(|x| x.id)
+    }
+    pub fn create_account_(&mut self, account_type: AccountType) -> Identity {
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:167]: nit: empty line

File: src/many-ledger/tests/common/mod.rs [hunk 14]
```diff
@@ -99,18 +118,65 @@ impl Setup {
...
+            many::server::module::ledger::SendArgs {
+                from: Some(from),
+                to,
+                amount: amount.into(),
+                symbol,
+            },
+        )?;
+        Ok(())
+    }
+    pub fn send_(&mut self, from: Identity, to: Identity, amount: impl Into<TokenAmount>) {
```
Comment (fmorency) [inline src/many-ledger/tests/common/mod.rs:158]: nit: empty line

**Review by fmorency — APPROVED** [review]
Very nice; thanks for all the helper functions! Loving it.

Minor comments/nits. 

Approved, assuming comments are addressed.
